controllers/dataplane_controller.go

```diff
@@ -0,0 +1,167 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	dataplane "github.com/kong/blixt/internal/dataplane/client"
+	"github.com/kong/blixt/pkg/vars"
+)
+
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gateways,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gateways/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=gateways/finalizers,verbs=update
+
+//+kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=services/status,verbs=get
+
+//+kubebuilder:rbac:groups=core,resources=endpoints,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=endpoints/status,verbs=get
+
+// DataplaneReconciler reconciles the dataplane pods.
+type DataplaneReconciler struct {
+	client.Client
+	scheme *runtime.Scheme
+
+	backendsClientManager *dataplane.BackendsClientManager
+
+	updates chan event.GenericEvent
+}
+
+func NewDataplaneReconciler(client client.Client, schema *runtime.Scheme, manager *dataplane.BackendsClientManager) *DataplaneReconciler {
+	return &DataplaneReconciler{
+		Client:                client,
+		scheme:                schema,
+		backendsClientManager: manager,
+		updates:               make(chan event.GenericEvent, 1),
+	}
+}
+
+var (
+	podOwnerKey = ".metadata.controller"
+	apiGVStr    = appsv1.SchemeGroupVersion.String()
+)
+
+// SetupWithManager loads the controller into the provided controller manager.
+func (r *DataplaneReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	// In order to allow our reconciler to quickly look up Pods by their owner, we’ll
+	// need an index. We declare an index key that we can later use with the client
+	// as a pseudo-field name, and then describe how to extract the indexed value from
+	// the Pod object. The indexer will automatically take care of namespaces for us,
+	// so we just have to extract the owner name if the Pod has a DaemonSet owner.
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Pod{}, podOwnerKey, func(rawObj client.Object) []string {
+		// grab the pod object, extract the owner...
+		pod := rawObj.(*corev1.Pod)
+		owner := metav1.GetControllerOf(pod)
+		if owner == nil {
+			return nil
+		}
+		// ...make sure it's a DaemonSet...
+		if owner.APIVersion != apiGVStr || owner.Kind != "DaemonSet" {
+			return nil
+		}
+
+		// ...and if so, return it
+		return []string{owner.Name}
+	}); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.DaemonSet{},
+			builder.WithPredicates(predicate.NewPredicateFuncs(r.daemonsetHasMatchingAnnotations)),
```

Comment:
Can you possibly filter this even more? i.e Only care about events where the DS's status changes since it's where we get the ready pods from anyways? 

```
status:
  currentNumberScheduled: 1
  desiredNumberScheduled: 1
  numberAvailable: 1
  numberMisscheduled: 0
  numberReady: 1
  observedGeneration: 1
  updatedNumberScheduled: 1
```

Reply:
I'm still relatively new to the tooling, did you mean something like that?
```go
	return ctrl.NewControllerManagedBy(mgr).
		For(&appsv1.DaemonSet{},
			builder.WithPredicates(predicate.NewPredicateFuncs(r.daemonsetHasMatchingAnnotations)),
		).
		WithEventFilter(predicate.Funcs{
			UpdateFunc: func(e event.UpdateEvent) bool {
				return true
			},
		}).
		Complete(r)
```

Reply:
Yeah close to that see https://sdk.operatorframework.io/docs/building-operators/golang/references/event-filtering/

In bpfd we do something like 

```
// Only reconcile if a bpfprogram has been created for the controller's program type.
func BpfProgramTypePredicate(kind string) predicate.Funcs {
	return predicate.Funcs{
		GenericFunc: func(e event.GenericEvent) bool {
			return e.Object.(*bpfdiov1alpha1.BpfProgram).Spec.Type == kind
		},
		CreateFunc: func(e event.CreateEvent) bool {
			return e.Object.(*bpfdiov1alpha1.BpfProgram).Spec.Type == kind
		},
		UpdateFunc: func(e event.UpdateEvent) bool {
			return e.ObjectNew.(*bpfdiov1alpha1.BpfProgram).Spec.Type == kind
		},
		DeleteFunc: func(e event.DeleteEvent) bool {
			return e.Object.(*bpfdiov1alpha1.BpfProgram).Spec.Type == kind
		},
	}
}
```

Reply:
(which is a bit verbose but has worked for us in the past)